app.py

```diff
@@ -707,6 +708,135 @@ def edit():
 			rows=rows, cols=100,
 			msg=msg)
 
+def prune_empty_non_terminals(tree: ParentedTree) -> ParentedTree:
+    # Recursively prune children first
+    for i in reversed(range(len(tree))):
+        child = tree[i]
+        if isinstance(child, ParentedTree):
+            pruned_child = prune_empty_non_terminals(child)
+            if len(pruned_child) == 0:
+                del tree[i]
+                
+    return tree
+
+def number_terminals(tree):
+	"""
+	Number the terminal nodes in a ParentedTree sequentially starting from 0.
+
+	Args:
+	tree (ParentedTree): The ParentedTree to renumber terminal nodes.
+
+	Returns:
+	ParentedTree: The updated ParentedTree with numbered terminal nodes.
+	"""
+	terminal_count = 0  # Initialize the terminal counter
+
+	def _number_terminals(node):
+		nonlocal terminal_count
+		if isinstance(node, ParentedTree):
+			for i, child in enumerate(node):
+				if isinstance(child, ParentedTree):
+					_number_terminals(child)
+				else:
+					# Assign a new terminal number
+					node[i] = terminal_count
+					terminal_count += 1
+
+	# Create a copy of the tree to avoid modifying the original
+	tree_copy = tree.copy(deep=True)
+	_number_terminals(tree_copy)
+	return tree_copy
+
+def remove_punctuation_nodes(tree):
+	"""
+	Recursively remove punctuation nodes from an NLTK ParentedTree.
+
+	Args:
+	tree (ParentedTree): The tree from which to remove punctuation nodes.
+
+	Returns:
+	ParentedTree: The tree with punctuation nodes removed.
+	"""
+	# Define a set of punctuation tags
+	punctuation_tags = {'.', ',', ':', '...', ';', '!', '?', '-', '--', '"', '``', "''", '(', ')', '[', ']', '{', '}', '-LRB-', '-RRB-', '-LSB-', '-RSB-', '-LCB-', '-RCB-'}
```

Comment:
This seems like it relates to the global definition of punctuation—should this be defined somewhere else to avoid incompatibilities?

Reply:
(Also, consider what preprocessing will be necessary for things like Unicode quotes/dashes. Maybe this happens somewhere else)

Reply:
Awesome, this is addressed in https://github.com/nschneid/activedop/pull/47/commits/80285bf669d29dfb9550039f6940b06f10bce436

Maybe Unicode quotes/dashes are replaced before tokenization (so that parser treats them the same as their non Unicode equivalents)?